Objects/listobject.c

```diff
@@ -1056,10 +1080,25 @@ list_inplace_repeat_lock_held(PyListObject *self, Py_ssize_t n)
     for (Py_ssize_t j = 0; j < input_size; j++) {
         _Py_RefcntAdd(items[j], n-1);
     }
-    // TODO: _Py_memory_repeat calls are not safe for shared lists in
-    // GIL_DISABLED builds. (See issue #129069)
+#ifndef Py_GIL_DISABLED
     _Py_memory_repeat((char *)items, sizeof(PyObject *)*output_size,
                       sizeof(PyObject *)*input_size);
+#else
+    if (_Py_IsOwnedByCurrentThread((PyObject *)self) && !_PyObject_GC_IS_SHARED(self)) {
+        // No other threads can read this list concurrently, so atomic repeat is not necessary.
+        _Py_memory_repeat((char *)items, sizeof(PyObject *)*output_size,
+                          sizeof(PyObject *)*input_size);
+        return 0;
+    }
+
+    PyObject **src = items;
+    PyObject **dst = items + input_size;
+    Py_ssize_t remaining = output_size - input_size;
+    while (remaining > 0) {
+        _Py_atomic_store_ptr_release(dst++, *src++);
+        remaining--;
+    }
```

Comment:
Why can't this just call list_atomic_memmove?

Reply:
Yeah, that will simplify this a bit and avoid the special case here (since `list_atomic_memmove` has it internally).